### DISCUSSION TITLE
Several class definitions for signal and and message

### DISCUSSION THREAD

**Discussion by ArseniSoitu** [cantools/cantools#549] [Q&A] [answered]
Hello.

I've found 2 different class definitions for **Signal**:
One in [**cantools/database/can/signal.py**](https://github.com/cantools/cantools/blob/master/cantools/database/can/signal.py#L152)
And another in [**cantools/database/can/c_source.py**](https://github.com/cantools/cantools/blob/master/cantools/database/can/c_source.py#L554)

Similar situation with class **Message**:
One in [**cantools/database/can/message.py**](https://github.com/cantools/cantools/blob/master/cantools/database/can/message.py#L42)
And another in [**cantools/database/can/c_source.py**](https://github.com/cantools/cantools/blob/master/cantools/database/can/c_source.py#L755)

Can you tell me what is the purpose to create different definitions?
Why not to use only one or create inheritance?

Thank you

**Comment by juleq** [comment, marked as answer]
One is part of cantools core business logic and the other is just providing methods to generate C code. If plugged in concerns like generating C code would be put into the core business logic, it would blow up in terms of size and maintenance.

**Reply by ArseniSoitu** [reply]
Thank you

But it looks like core logic and C generator logic both operate with entries from database.
And the database consists from messages, and those messages consist from signals.

Or am I missing something?

**Reply by andlaus** [reply]
as @juleq said: the messages and signals in the database are represented by objects of the classes from `cantools/database/can/{message,signal}.py` while `cantools/database/can/c_source/{message,signal}.py` uses these objects to produce C source code for messages and signals. What's probably confusing is that the latter are located in `cantools/database/can/c_source/` although they are IMO not at all related to the database representation...
